### PR TITLE
Format Parametrization for Test Infrastructure & Bug Fixes

### PR DESCRIPTION
**Pull request by skotaracTT**
### Format Parametrization for Test Infrastructure
* **Feature Addition**: A new feature has been added to parameterize test formats, providing the flexibility to select and combine various formats during testing. This enhancement enables users to configure and test five distinct formats: unpacker_src, unpacker_dst, math, packer_src, and packer_dst. Each format is specifically associated with one of the system components: Unpacker, Math, and Packer LLKs (Low-Level Kernels).
* **Code Cleanup and Refactor**: As part of this update, the Python test code has been refactored to minimize repetition, improve readability, and enhance the clarity of the setup process for defining test parameters. This makes the test code more maintainable, reducing the chances of errors and making it easier for users to understand how to set up their test configurations.
* **Dynamic Format Combinations**: With the introduction of the generate_format_combinations function, users can now specify which formats they wish to test. This function automatically generates all combinations of the selected formats, allowing users to test a wide variety of scenarios. This is particularly useful for format sweep testing, where the goal is to exhaustively test all combinations of formats to identify potential edge cases and weaknesses in the system.
* **Testing Benefits**: By enabling the generation of format combinations, this feature allows users to run tests across all potential format permutations. The result is a comprehensive testing approach that can:
    * Identify passing combinations: Ensures that the system works as expected for the chosen format combinations.
    * Highlight expected failures: Helps in confirming that known issues occur under the correct conditions, providing a clearer understanding of what is expected to fail.
    * Spot unexpected failures: Enables the detection of failures that may arise from unanticipated interactions between formats, helping to uncover hidden bugs or inconsistencies.
Overall, this approach provides a more thorough and systematic way to validate the system, enabling better test coverage and more insightful debugging when failures occur.
* **Bug Fixes** 
    * Bug in test_pack_untilize.py was not calling llk functions properly for BH, that is why tests on main pass for WH but fail for BH. Fixed issue in `pack_untilize_test.cpp` so that they pass on both WH and BH.
    * Located bugs in the test infrastructure, specifically in **`unpack.py`**. One issue was related to the conversion of bfp8_b into float16 format in the bfp8_to_float_block() function. Another problem was found in the order of reading bytes from the output in L1 in the functions unpack_bfp16, unpack_fp16_b, and unpack_bfp8_b. The issue arises when the input format differs from the output format, causing discrepancies in byte reading order, which has now been addressed.
* **Format Configuration Management:** Added a new param_config.py file to handle the format configurations in the background, providing a clean and easy-to-use interface in the Python test code. The new FormatConfig class (found in param_config.py) stores the available formats in an organized and centralized manner. This class mimics a structure in C, making it easier to manage and extend the format configuration in the future. The five formats are listed in the formats attribute of the class, which can be accessed programmatically when setting up tests.

### Why This Is Important:
* Improved Test Coverage: By automating the generation of format combinations, we ensure that all possible combinations of formats are tested without the need for manual configuration. This increases the test coverage significantly.
* Easier Test Setup: The FormatConfig class and the generate_format_combinations function simplify the process of defining and running tests with different format configurations. Users no longer need to manually set each format combination, which saves time and reduces the chances of human error.
* Flexibility for Future Updates: This change sets the stage for adding new formats in the future without having to overhaul the testing infrastructure. By simply extending the FormatConfig class with new formats, users can quickly integrate new format combinations into their tests.
* Better Debugging: The ability to see which format combinations pass or fail gives developers better insights into where the system may be vulnerable or prone to failure, leading to faster identification and resolution of bugs.
* Resolved Bugs: Fixing the issues in unpack.py ensures more accurate handling of format conversions and byte reading during unpacking, improving the stability and reliability of the unpacking operations when dealing with different input and output formats.

### Issues to Report
* Tests on main branch tend to fail for `test_multiple_tiles_eltwise.py` only on BH but on WH they pass. When running on BH the mailbox assertions in `assert_Tensix_operations_finished` fail.
* Tests skipping is not aligned for BH and WH architecture. For `test_eltwise_unary_sfpu.py` skips if `format == "Float16" and dest_acc == "DEST_ACC"`, this works fine for WH because the case is not supported, however, this case passes on BH, and instead BH test fails when `format == "Float16" and dest_acc == "" so this should be skipped for BH.
* To test full format sweep, you will be running hundreds of tests. Tests that fail do not reset the board afterwards, so they may cause incorrect behavior in subsequent passing tests. To ensure accurate results, the board must be reset after each test.

### CODE REVIEW

File: tests/python_tests/helpers/format_arg_mapping.py [hunk 1]
```diff
@@ -30,6 +30,46 @@
     "reduce_scalar": "REDUCE_SCALAR_OPERATION",
 }
 
+unpack_src_dict = {
+    "Float32": "UNPACK_SRC_FLOAT32",
```
Comment (fvranicTT) [inline tests/python_tests/helpers/format_arg_mapping.py:34]: We should avoid using string comparison in too many places. Strings are error prone and easy to mistype, and those errors could be hard to catch later. Instead, it'd be great if we could use enum class:
```python
class DataFormat(Enum):
    Float16 = "Float16"
    Float16_B = "Flaot16_b"
    Bfp8_B = "Bfp8_b"
    Float32 = "Float32"
    Int32 = "Int32"
```
And we only need to check strings once. It's easier to spot I made a typo in "Flaot16_b" once, then finding more instances of typos throughout the code.

File: tests/python_tests/helpers/unpack.py [hunk 2]
```diff
@@ -100,8 +138,18 @@ def unpack_bfp8_b(bfp8_block, sfpu=False):
...
+    # Patch Up! Fixes incorrect reading of numbers in L1:
+    # When input source i.e `unpack_src` is not BFP8_B, but it is packed as BFP8_B then consecutive pairs of numbers are inverted and placed in L1.
+    # Instead of being placed as (a,b,c,d,e,f,...) in L1, they are placed as (b,a,d,c,f,e,...).
+    # This caused the test to fail as the results were correctly computed but read from L1 incorrectly.
+    # The loop reinverts the numbers back to their correct positions in order to read them properly and pass the test as expected.
+    if unpack_src != pack_dst:
+        for i in range(0, len(bfloat16_values), 2):
+            tmp = bfloat16_values[i]
+            bfloat16_values[i] = bfloat16_values[i + 1]
+            bfloat16_values[i + 1] = tmp
```
Comment (fvranicTT) [inline tests/python_tests/helpers/unpack.py:153]:
```suggestion
            bfloat16_values[i], bfloat16_values[i + 1] = bfloat16_values[i + 1], bfloat16_values[i]
```

File: tests/python_tests/helpers/unpack.py [hunk 3]
```diff
@@ -7,36 +7,62 @@
...
+    # Patch Up! Fixes incorrect reading of numbers in L1:
+    # When input format i.e `unpack_src` is BFP8_b but the result is packed into a different format then consecutive pairs of numbers are inverted in L1.
+    # Instead of being placed as (a,b,c,d,e,f,...) in L1, they are placed as (b,a,d,c,f,e,...).
+    # This caused the test to fail as the results were correctly computed but read incorrectly.
+    # The loop reinverts the numbers back to their correct positions in order to read them properly and pass the test as expected.
+    if unpack_src == "Bfp8_b" and pack_dst != unpack_src:
+        for i in range(0, len(ret), 2):
+            tmp = ret[i]
+            ret[i] = ret[i + 1]
+            ret[i + 1] = tmp
```
Comment (fvranicTT) [inline tests/python_tests/helpers/unpack.py:63]:
```suggestion
            ret[i], ret[i + 1] = ret[i + 1], ret[i]
```

File: tests/python_tests/helpers/unpack.py [hunk 4]
```diff
@@ -7,36 +7,62 @@
...
```
Comment (fvranicTT) [inline tests/python_tests/helpers/unpack.py:38]:
```suggestion
            ret[i], ret[i + 1] = ret[i + 1], ret[i]
```

File: tests/sources/eltwise_binary_test.cc
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/sources/eltwise_binary_test.cc:1]: Should be removed. It is covered multiple_tiles test when tile_cnt = 1

File: tests/python_tests/helpers/device.py [hunk 5]
```diff
@@ -84,18 +90,25 @@ def get_result_from_device(
...
+            )  # Bug patchup in (unpack.py): in case unpack_src is Bfp8_b != pack_dst, L1 must be read in different order to extract correct results
+        else:
+            return unpack_func(read_data_bytes)
     else:
-        raise ValueError(f"Unsupported format: {format}")
+        raise ValueError(f"Unsupported format: {formats.pack_dst}")
 
 
 def assert_tensix_operations_finished(core_loc: str = "0,0"):
+    # time.sleep(1) # assertions below will incorrectly fail if run immediately after the ELF files are run, this needs to be fixed
```
Comment (ldjurovicTT) [inline tests/python_tests/helpers/device.py:111]: What does it mean incorrectly fail? Maybe add a bit more details about this

File: tests/python_tests/helpers/format_arg_mapping.py
Comment (ldjurovicTT) [inline tests/python_tests/helpers/format_arg_mapping.py:1]: Do We still need format_dict?

File: tests/python_tests/test_eltwise_unary_datacopy.py [hunk 6]
```diff
@@ -10,37 +11,42 @@ def generate_golden(operand1, format):
...
-    (format, dest_acc, testname)
-    for format in ["Float32", "Bfp8_b", "Float16_b", "Float16", "Int32"]
-    for dest_acc in ["", "DEST_ACC"]
-    for testname in ["eltwise_unary_datacopy_test"]
-]
-
-param_ids = [f" format={comb[0]} | dest_acc={comb[1]} " for comb in param_combinations]
+full_sweep = False
+all_format_combos = generate_format_combinations(
+    ["Float32", "Bfp8_b", "Float16_b", "Float16", "Int32"], not full_sweep
```
Comment (fvranicTT) [inline tests/python_tests/test_eltwise_unary_datacopy.py:16]:
```suggestion
    formats = ["Float32", "Bfp8_b", "Float16_b", "Float16", "Int32"], all_same = True
```
This looks easier to understand to the untrained eye.

File: tests/helpers/include/params.h [hunk 7]
```diff
@@ -27,20 +27,84 @@ constexpr std::underlying_type_t<DataFormat> get_data_format(DataFormat format)
...
+#ifdef UNPACK_SRC_INT32
+constexpr auto UNPACK_IN = get_data_format(DataFormat::Int32);
 #endif
-#ifdef FORMAT_BFP8_B
-constexpr auto DATA_FORMAT = get_data_format(DataFormat::Bfp8_b);
+#ifdef UNPACK_SRC_BFP8_B
+constexpr auto UNPACK_IN = get_data_format(DataFormat::Bfp8_b);
+#endif
+
+#ifdef UNPACK_DST_FLOAT16_B
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/helpers/include/params.h:46]: Yes I agree! Especially since we will be adding more format coverage, this would be great to minimize repetition and reduce error-prone code.

File: tests/python_tests/helpers/param_config.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    """
+
+    unpack_src: str
+    unpack_dst: str
+    math: str
+    pack_src: str
+    pack_dst: str
+
+
+included_params = []
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:34]: yes we can, I will make a another function that will store this variable and call the function which uses included_params.

File: tests/python_tests/helpers/unpack.py [hunk 9]
```diff
@@ -7,36 +7,62 @@
 import torch
 from .utils import reverse_endian_chunk
 
+unpacked_bfp8 = {}
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/unpack.py:10]: Yes! same solution as above proposed.

File: tests/python_tests/helpers/param_config.py
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:1]: The file is used to simplify our test parameter configurations, so it's not repeated in every test file. It configures formats and parametrizes our test input (parameters) which its why I named it as such because it configures our test parameters. I debated putting our format configuration in a separate file but for now it doesn't need much code, so I decided not to add another file to the cluster.

File: tests/python_tests/helpers/format_arg_mapping.py (hunk 1 above)
Comment (skotaracTT) [reply tests/python_tests/helpers/format_arg_mapping.py:34]: This is great! Absolutely agree, will make the changes necessary.

File: tests/sources/eltwise_binary_test.cc [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    _llk_pack_init_<false, false, DstTileFaceLayout::RowMajor, false>(PACK_OUT);
+
+#ifdef ARCH_BLACKHOLE
+    _llk_pack_dest_init_<DstSync::SyncFull, DstTileFaceLayout::RowMajor, is_fp32_dest_acc_en>();
+#else
+    _llk_pack_dest_init_<DstSync::SyncFull, DstTileFaceLayout::RowMajor, false, false>();
+#endif
+
+    _llk_packer_wait_for_math_done_();
+    _llk_pack_<DstSync::SyncFull, false, is_fp32_dest_acc_en>(0, (std::uint32_t)buffer_Dest / 16 - 1);
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/sources/eltwise_binary_test.cc:85]: By accidentally forgot to delete this file, we are no longer testing since this test falls into `test_multiple_tiles_eltwise`

File: tests/sources/eltwise_binary_test.cc [hunk 11]
```diff
@@ -0,0 +1,89 @@
+#include <cstdint>
+#include <cstdio>
+
+#include "ckernel.h"
+#include "llk_defs.h"
+
+// Globals
+uint32_t unp_cfg_context        = 0;
+uint32_t pack_sync_tile_dst_ptr = 0;
+volatile uint32_t tt_l1_ptr l1_buffer[16] __attribute__((section(".text#"))) __attribute__((aligned(16)));
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/sources/eltwise_binary_test.cc:10]: This file is to be deleted, we are no longer testing since this test falls into `test_multiple_tiles_eltwise`

File: tests/python_tests/helpers/unpack.py [hunk 12]
```diff
@@ -7,36 +7,62 @@
...
+    # When input format i.e `unpack_src` is BFP8_b but the result is packed into a different format then consecutive pairs of numbers are inverted in L1.
+    # Instead of being placed as (a,b,c,d,e,f,...) in L1, they are placed as (b,a,d,c,f,e,...).
+    # This caused the test to fail as the results were correctly computed but read incorrectly.
+    # The loop reinverts the numbers back to their correct positions in order to read them properly and pass the test as expected.
+    if unpack_src == "Bfp8_b" and pack_dst != unpack_src:
+        for i in range(0, len(ret), 2):
+            tmp = ret[i]
+            ret[i] = ret[i + 1]
+            ret[i + 1] = tmp
+    return ret
```
Comment (ldjurovicTT) [inline tests/python_tests/helpers/unpack.py:39]: Is conversion from fp16 to bfp8_b even allowed? I think that there is now HW supported conversion from A to B formats. Correct me if I'm wrong

File: tests/python_tests/test_eltwise_unary_datacopy.py [hunk 13]
```diff
@@ -10,37 +11,42 @@ def generate_golden(operand1, format):
...
             reason="Skipping test for 32 bit wide data without 32 bit accumulation in Dest"
         )
 
-    src_A, src_B = generate_stimuli(format)
+    #  When running hundreds of tests, failing tests may cause incorrect behavior in subsequent passing tests.
+    #  To ensure accurate results, for now we reset board after each test.
+    #  Fix this: so we only reset after failing tests
+    if full_sweep:
+        run_shell_command(f"cd .. && make clean")
+        run_shell_command(f"tt-smi -r 0")
```
Comment (ldjurovicTT) [inline tests/python_tests/test_eltwise_unary_datacopy.py:41]: This here is a very important topic. Let's try figuring out how to track if this happens. Maybe We can run tests with out run_test.sh ( when it finally works) and from there conclude if We should reset or not. Even outside python script could be of help

File: tests/helpers/include/params.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (fvranicTT) [reply tests/helpers/include/params.h:46]: Maybe to be on the safe side:
```cpp
#if defined(UNPACK_DST_FLOAT16_B)
UNPACK_CASE(Float16_b)
#elif defined(UNPACK_DST_FLOAT16)
UNPACK_CASE(Float16)
#elif defined(UNPACK_DST_FLOAT32)
UNPACK_CASE(Float32)
#elif defined(UNPACK_DST_INT32)
UNPACK_CASE(Int32)
#else
UNPACK_CASE(Bfp8_b)
#endif
```

File: tests/python_tests/test_eltwise_unary_datacopy.py (hunk 13 above)
Comment (skotaracTT) [reply tests/python_tests/test_eltwise_unary_datacopy.py:41]: Yes I agree! This here is a temporary fix, was done in order to fully run tests and in order to see how they truly behave without being affected by failing tests. Will be adjusted, sounds good!

File: tests/python_tests/test_eltwise_unary_datacopy.py (hunk 6 above)
Comment (skotaracTT) [reply tests/python_tests/test_eltwise_unary_datacopy.py:16]: Agreed! Also follows function parameter construct better

File: tests/python_tests/helpers/unpack.py (hunk 12 above)
Comment (ldjurovicTT) [reply tests/python_tests/helpers/unpack.py:39]: In that case great!

File: tests/sources/eltwise_binary_test.cc [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+#endif
+
+#ifdef LLK_TRISC_UNPACK
+#undef MULTIPLE_OPS
+
+#include "llk_unpack_AB.h"
+#include "llk_unpack_common.h"
+#include "params.h"
+
+volatile uint32_t* buffer_A = (volatile uint32_t*)0x1a000;
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/sources/eltwise_binary_test.cc]: This file is deleted and outdated, we are no longer testing since this test falls into test_multiple_tiles_eltwise

File: tests/sources/eltwise_binary_test.cc [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+#endif
+
+#ifdef LLK_TRISC_PACK
+
+#include "llk_pack.h"
+#include "llk_pack_common.h"
+#include "params.h"
+
+volatile uint32_t* buffer_Dest = (volatile uint32_t*)0x1c000;
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/sources/eltwise_binary_test.cc]: This file is deleted and outdated, we are no longer testing since this test falls into test_multiple_tiles_eltwise

File: tests/sources/eltwise_binary_test.cc [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+#include "params.h"
+
+volatile uint32_t* buffer_A = (volatile uint32_t*)0x1a000;
+volatile uint32_t* buffer_B = (volatile uint32_t*)0x1b000;
+
+void run_kernel()
+{
+    _llk_unpack_AB_hw_configure_<is_fp32_dest_acc_en, StochRndType::None>(UNPACK_IN, UNPACK_IN, UNPACK_OUT, UNPACK_OUT);
+    _llk_unpack_AB_init_<>();
+    _llk_unpack_AB_<>((std::uint32_t)buffer_A / 16 - 1, (std::uint32_t)buffer_B / 16 - 1);
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/sources/eltwise_binary_test.cc]: This file is deleted and outdated, we are no longer testing since this test falls into test_multiple_tiles_eltwise

File: tests/python_tests/helpers/format_arg_mapping.py
Comment (skotaracTT) [reply tests/python_tests/helpers/format_arg_mapping.py:1]: No, I will remove it.

File: tests/helpers/include/params.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/helpers/include/params.h:46]: Yes, even better.

File: tests/python_tests/helpers/format_arg_mapping.py
Comment (fvranicTT) [reply tests/python_tests/helpers/format_arg_mapping.py:1]: Don’t you need it for specifying data type when creating tensors?

Comment (skotaracTT) [reply tests/python_tests/helpers/format_arg_mapping.py:1]: However, I thought I read `format_args_dict`, which is no longer needed. It was used before when testing just one format, but now we have other dictionaries for testing specific formats for unpacker, pack and math.

File: tests/python_tests/helpers/unpack.py [hunk 17]
```diff
@@ -7,36 +7,62 @@
...
-def unpack_bfp16(packed_list):
+def unpack_bfp16(packed_list, unpack_src, pack_dst):
     def bytes_to_bfloat16(byte_list):
         bytes_data = bytes(byte_list[:2] + [0, 0])  # Ensure we include padding
         unpacked_value = struct.unpack(">f", bytes_data)[0]
         return unpacked_value
 
     limited_packed_list = packed_list[:2048]
-    return [
+    ret = [
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/unpack.py]: Done ✅

File: tests/python_tests/helpers/format_arg_mapping.py (hunk 1 above)
Comment (skotaracTT) [reply tests/python_tests/helpers/format_arg_mapping.py]: Done ✅

File: tests/python_tests/helpers/unpack.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/unpack.py]: Done ✅

File: tests/python_tests/helpers/param_config.py
[the remark this answers is not in the document]
Comment (fvranicTT) [reply tests/python_tests/helpers/param_config.py:1]: Yes, but I thought it would be more suitable to name it format_config.py, since the class is named FormatConfig.

File: tests/python_tests/helpers/format_arg_mapping.py
Comment (fvranicTT) [reply tests/python_tests/helpers/format_arg_mapping.py:1]: So we can then safely remove it 🙃

Comment (skotaracTT) [reply tests/python_tests/helpers/format_arg_mapping.py:1]: Done🫡

File: tests/python_tests/helpers/param_config.py
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:1]: Yes, I agree with you here. Will be much easier for users to locate definition. I have added a new file called `format_config` which holds classes used to create and define Formats.✅

File: tests/helpers/include/params.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/helpers/include/params.h:52]: Done ✅ Will look into making this shorter in the future (less repetition).

File: tests/python_tests/helpers/format_config.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+    Example:
+    >>> formats = FormatConfig(unpack_src="Float32", unpack_dst="Float16", math="add", pack_src="Float16", pack_dst="Float32")
+    >>> print(formats.unpack_src)
+    Float32
+    >>> print(formats.pack_src)
+    Float16
+    """
+
+    unpack_src: str
```
Comment (fvranicTT) [inline tests/python_tests/helpers/format_config.py:28]: Can these now be of type DataFormat?

Comment (skotaracTT) [reply tests/python_tests/helpers/format_config.py:28]: Done ✅

File: tests/python_tests/helpers/device.py [hunk 19]
```diff
@@ -73,30 +81,54 @@ def write_stimuli_to_l1(buffer_A, buffer_B, stimuli_format, core_loc="0,0", tile
...
+            return unpack_func(read_data_bytes)
     else:
-        raise ValueError(f"Unsupported format: {format}")
+        raise ValueError(f"Unsupported format: {formats.pack_dst}")
+
+
+def check_value_with_timeout(core_loc, mailbox_addr, timeout=0, poll_interval=0.1):
+    start_time = time.time()
+
+    while time.time() - start_time < timeout:
```
Comment (fvranicTT) [inline tests/python_tests/helpers/device.py:118]:
```suggestion
    end_time = time.time() + timeout

    while time.time() < end_time:
```
Seems easier to read.

File: tests/python_tests/helpers/device.py [hunk 20]
```diff
@@ -73,30 +81,54 @@ def write_stimuli_to_l1(buffer_A, buffer_B, stimuli_format, core_loc="0,0", tile
...
+                read_data_bytes, formats.unpack_src, formats.pack_dst
+            )  # Bug patchup in (unpack.py): in case unpack_src is Bfp8_b != pack_dst, L1 must be read in different order to extract correct results
+        else:
+            return unpack_func(read_data_bytes)
     else:
-        raise ValueError(f"Unsupported format: {format}")
+        raise ValueError(f"Unsupported format: {formats.pack_dst}")
+
+
+def check_value_with_timeout(core_loc, mailbox_addr, timeout=0, poll_interval=0.1):
```
Comment (fvranicTT) [inline tests/python_tests/helpers/device.py:115]:
```suggestion
def assert_value_with_timeout(core_loc, mailbox_addr, timeout=0, poll_interval=0.1):
```

File: tests/python_tests/test_eltwise_unary_sfpu.py [hunk 21]
```diff
@@ -23,44 +23,57 @@ def generate_golden(operation, operand1, data_format):
...
+    if (
+        formats.unpack_src in [DataFormat.Float32, DataFormat.Int32]
+        and dest_acc != "DEST_ACC"
+    ):
         pytest.skip(
             reason="Skipping test for 32 bit wide data without 32 bit accumulation in Dest"
         )
-
-    if format == "Float16" and dest_acc == "DEST_ACC":
+    if (
```
Comment (fvranicTT) [inline tests/python_tests/test_eltwise_unary_sfpu.py:53]: But is this really the case? So far, we've been skipping this test if format was Float16 and dest_acc is enabled.
In the upcoming change, we only check against blackhole, i.e. wormhole would run this test and fail, correct?

Comment (fvranicTT) [reply tests/python_tests/test_eltwise_unary_sfpu.py:53]:
```python
if (formats.unpack_src == DataFormat.Float16 and ((get_chip_architecture() == "blackhole")
        or (get_chip_architecture() == "wormhole" and dest_acc == "DEST_ACC"))
):
```

Comment (skotaracTT) [reply tests/python_tests/test_eltwise_unary_sfpu.py:53]: My bad, that is a typo. The test format  fails on bh when dest accumulation is off, and so it should be skipped. However, this passes on wh, but when dest accumulation is on then wh fails but passes for BH. So we skip on wh when dest_acc = "DEST_ACC" and skip on bh when dest_acc =  "".

Comment (fvranicTT) [reply tests/python_tests/test_eltwise_unary_sfpu.py]: Ok. I’ convinced.

**Review by fvranicTT — APPROVED** [review]

**Comment by fvranicTT** [comment]
🚀

**Comment by skotaracTT** [comment]
🤝🤩
